### ISSUE TITLE
postcss is injecting generated .css file into source list

### ISSUE THREAD

**Issue by airdrummingfool** [postcss/postcss#907]
Using gulp and gulp-usemin, I compile a set of LESS files into one generated `main.css` file. When including postcss (without any plugins) in the pipeline, the sourcemaps include this generated `main.css` file in the `sources` list.

Simplified version of `main.css.map` file without postcss:

```
{
"version":3,
"sources":["app/styles/less/common.less","app/styles/less/custom.less","app/styles/less/login.less","app/styles/less/admin.less"],
"names":[],
"mappings":"...",
"file":"main.css",
"sourcesContent":["..."]
}
```

Simplified version of `main.css.map` with postcss (without any plugins):

```
{
"version":3,
"sources":["app/styles/less/common.less","styles/main.css","app/styles/less/custom.less","app/styles/less/login.less","app/styles/less/admin.less"],
"names":[],
"mappings":"...",
"file":"main.css",
"sourcesContent":["..."]
}
```

How can I prevent postcss from adding `"styles/main.css"` to the sourcemap sources? I've tried various permutations of the `from`, `to`, and `map` options listed on the [postcss and sourcemaps page](https://github.com/postcss/postcss/blob/master/docs/source-maps.md), but I can't figure it out.

**Comment by ai** [comment]
I think problem is that not all symbols has mappings in Less source map (for example, whitespaces). So when PostCSS create mappings for all symbols, some of them connect with Less’s mapping, but some have no connections.

Honestly, I don’t know best solution right here. And do we really need to spend time to fix it? ;)

If you really want to fix it, you can use [source-map-visualization](https://sokra.github.io/source-map-visualization/) to see differences between PostCSS and Less mappings. If this mappings was really strange feel free to open new issue or send a PR.

**Comment by erquhart** [comment]
> And do we really need to spend time to fix it? ;)

The resulting sourcemap causes the browser to look for a file that doesn't exist, and bad data is a problem when anyone wants to use the postcss api for anything serious, like any kind of wrapper tool.

In my case, I'm copying the sources files to dest, and have no reliable way to determine which entry in the sources array is the bogus one. PostCSS reliably adds it and shouldn't.

Does this at least merit remaining open? I'm currently blocked by this and am unsure how to proceed. Maybe if it's open someone will eventually have time to look into it.

**Comment by ai** [comment]
@erquhart I think you should not always be sure in `sources`. What if some other tool will miss mapping for some `a{}`? In this case PostCSS anyway will use `.css` file mapping.

Sometimes PostCSS plugins developers write bad code and miss `Node#input`. In this case PostCSS set `<no source>` source (it helps to find this mistakes).

So you should write:

```js
if (fs.exists(source)) {
  doMagic(source)
}
```

**Comment by ai** [comment]
Of course, if you have time for investigation I will accept PR.

As I told you should start from `source-map-visualization` to see differences between PostCSS and Less mappings. Maybe you will find, that Less miss mappings for some symbols or whitespaces. Or maybe it could be a small PostCSS mistake.

**Comment by erquhart** [comment]
@ai right, but what if a source's path is bad for some other reason? I'd like to warn the user and maybe even error out. But due to this issue, my only route currently is to silently fail as you're suggesting. I'm not able to improve the user experience by providing useful warnings that may be meaningful.

**Comment by ai** [comment]
@erquhart it is good that you try to provide a better UI for users :).

OK, show me a that strange mappings (you can see it by `source-map-visualization`). Then I will tell you how we can fix it.

**Comment by erquhart** [comment]
@ai here ya go:

[source -> transpiled](https://sokra.github.io/source-map-visualization/#base64,LmZvbyB7CiAgYmFja2dyb3VuZDogIzAwMDsKfQpib2R5IHsKICBjb2xvcjogYmxhY2s7Cn0KYm9keSB7CiAgY29sb3I6IHdoaXRlOwp9,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInNyYy90ZXN0L2ltcG9ydHMvaW1wb3J0Lmxlc3MiLCJzcmMvaW1wb3J0Lmxlc3MiLCJzcmMvdGVzdC9sZXNzLmxlc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7RUFDRSxnQkFBQTs7QUNERjtFQUNFLFlBQUE7O0FDS0Y7RUFDRSxZQUFBIiwiZmlsZSI6ImV4YW1wbGUuanMifQ==,LmZvbyB7CiAgYmFja2dyb3VuZDogQGJsYWNrOwp9Cg==,Ym9keSB7CiAgY29sb3I6IGJsYWNrOwp9,QGJsYWNrOiAjMDAwOwoKQGltcG9ydCAiaW1wb3J0cy9pbXBvcnQiOwoKQGltcG9ydCAiLi4vaW1wb3J0IjsKCmJvZHkgewogIGNvbG9yOiB3aGl0ZTsKfQo=)

[transpiled -> minified](https://sokra.github.io/source-map-visualization/#base64,LmZvb3tiYWNrZ3JvdW5kOiMwMDB9Ym9keXtjb2xvcjojMDAwO2NvbG9yOiNmZmZ9,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInNyYy90ZXN0L2ltcG9ydHMvaW1wb3J0Lmxlc3MiLCJsZXNzLmxlc3MiLCJzcmMvaW1wb3J0Lmxlc3MiLCJzcmMvdGVzdC9sZXNzLmxlc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUEsS0FDRSxlQUFBLENDQ0QsQUNGRCxLQUNFLFdBQUEsQUNNQSxVQUFBLENGRkQiLCJmaWxlIjoibGVzcy5sZXNzIn0=,LmZvbyB7CiAgYmFja2dyb3VuZDogQGJsYWNrOwp9Cg==,QGJsYWNrOiAjMDAwOwoKQGltcG9ydCAiaW1wb3J0cy9pbXBvcnQiOwoKQGltcG9ydCAiLi4vaW1wb3J0IjsKCmJvZHkgewogIGNvbG9yOiB3aGl0ZTsKfQo=,Ym9keSB7CiAgY29sb3I6IGJsYWNrOwp9,QGJsYWNrOiAjMDAwOwoKQGltcG9ydCAiaW1wb3J0cy9pbXBvcnQiOwoKQGltcG9ydCAiLi4vaW1wb3J0IjsKCmJvZHkgewogIGNvbG9yOiB3aGl0ZTsKfQo=)

**Comment by ai** [comment]
Strange. I didn’t see a `.css` mapping there O_o. Maybe we don’t clean `sources`?

Could you show `result.map._mappings`? (set `map: { inline: false }` to get `result.map`)

**Comment by erquhart** [comment]
That is the map from result. I don't use inline maps.

**Comment by ai** [comment]
@erquhart sure, just show me a `._mappings` inside this `result.map`.

**Comment by erquhart** [comment]
Ah, didn't even realize the map itself isn't provided in the visualization. Below is the entire map produced by postCSS, prettified. Let me know if you need the input map as well.

```
{
  "version": 3,
  "sources": [
    "src/test/imports/import.less",
    "less.less",
    "src/import.less",
    "src/test/less.less"
  ],
  "names": [],
  "mappings": "AAAA,KACE,eAAA,CCCD,ACFD,KACE,WAAA,ACMA,UAAA,CFFD",
  "file": "less.less"
}
```

**Comment by ai** [comment]
I need a PostCSS map mappings. Not a PostCSS map JSON, but `result.map._mappings`.

`result` is that thing from `postcss(plugins).process(css) //=> result` (when you use low-level JS API). `_mappings` is a private method from `sourceMap.SourceMapGenerator`, which show decrypted mapping.

(Or just create a simple test project, I could look at it on next weekend.)

**Comment by erquhart** [comment]
Gotcha:

```
MappingList {
  _array:
   [ { generatedLine: 1,
       generatedColumn: 0,
       originalLine: 1,
       originalColumn: 0,
       source: 'src/test/imports/import.less',
       name: null },
     { generatedLine: 1,
       generatedColumn: 5,
       originalLine: 2,
       originalColumn: 2,
       source: 'src/test/imports/import.less',
       name: null },
     { generatedLine: 1,
       generatedColumn: 20,
       originalLine: 2,
       originalColumn: 2,
       source: 'src/test/imports/import.less',
       name: null },
     { generatedLine: 1,
       generatedColumn: 21,
       originalLine: 3,
       originalColumn: 1,
       source: 'less.less',
       name: null },
     { generatedLine: 1,
       generatedColumn: 21,
       originalLine: 1,
       originalColumn: 0,
       source: 'src/import.less',
       name: null },
     { generatedLine: 1,
       generatedColumn: 26,
       originalLine: 2,
       originalColumn: 2,
       source: 'src/import.less',
       name: null },
     { generatedLine: 1,
       generatedColumn: 37,
       originalLine: 2,
       originalColumn: 2,
       source: 'src/import.less',
       name: null },
     { generatedLine: 1,
       generatedColumn: 37,
       originalLine: 8,
       originalColumn: 2,
       source: 'src/test/less.less',
       name: null },
     { generatedLine: 1,
       generatedColumn: 47,
       originalLine: 8,
       originalColumn: 2,
       source: 'src/test/less.less',
       name: null },
     { generatedLine: 1,
       generatedColumn: 48,
       originalLine: 6,
       originalColumn: 1,
       source: 'less.less',
       name: null } ],
  _sorted: true,
  _last:
   { generatedLine: 1,
     generatedColumn: 48,
     originalLine: 6,
     originalColumn: 1,
     source: 'less.less',
     name: null } }
```

**Comment by ai** [comment]
Sure, seems like we have a PostCSS issue here. `sources` are not cleaned.

**Comment by erquhart** [comment]
Cool. Looks like PostCSS is mapping the closing brackets?

**Comment by ai** [comment]
I need to go to bed right now. Could you create a small test project?

**Comment by erquhart** [comment]
I'll get one together soon. Thanks for your help, have a great night.

**Comment by airdrummingfool** [comment]
Here's a simple example project: https://github.com/airdrummingfool/postcss-issue-907

If you remove the empty postcss task (`postcss([])`) from the usemin `less` build pipeline, `styles/main.css` does not show up in the `sources`.

**Comment by ai** [comment]
I checked PostCSS sources. Main problem, PostCSS doesn’t generate `sources`. SO there is no way to fix it without dangerous hack.

Try to open issue in `source-map`: https://github.com/mozilla/source-map

By some reason they didn’t clean `sources`.

**Comment by airdrummingfool** [comment]
@ai I'm not sure I understand... without PostCSS, the issue does not happen. Can you elaborate on how it is a bug in `source-map` instead of PostCSS?

**Comment by ai** [comment]
@airdrummingfool because PostCSS doesn’t update Sass source map. PostCSS only generates new source map.

`source-map` combines two source maps.
